urlaa051fbd3d_WheezePuppet_collegeDiversity_tgz-pJ8-edu_umw_cpsc_collegediversity_SimJ8/src/edu/umw/cpsc/collegediversity/FreshmanHousingSelection.java

```diff
@@ -0,0 +1,214 @@
+package edu.umw.cpsc.collegediversity;
+
+import java.util.*;
+import sim.util.*;
+import ec.util.*;
+
+/**
+ * A Singleton class that can {@link #assign} a set of freshman {@link
+ * Student}s to their freshman dorms. 
+ * @see UpperclassHousingSelection
+ */
+public class FreshmanHousingSelection{
+
+    private static FreshmanHousingSelection theInstance;
+    private Bag dorms;
+
+    private FreshmanHousingSelection(){
+
+        dorms = new Bag();
+
+        Dorm Alvey = new Dorm("Alvey", false, true, 73);
+        Dorm Bushnell = new Dorm("Bushnell", false, true, 76);
+        Dorm Jefferson = new Dorm("Jefferson", false, true, 96);
+        Dorm Randolph = new Dorm("Randolph", false, true, 93);
+        Dorm Russell = new Dorm("Russell", false, true, 87);
+
+        //(actually mixed-year but for simplicity I'm keeping it freshman 
+        // only for now)
+        Dorm Virginia = new Dorm("Virginia", true, true, 92); 
+
+        dorms.add(Virginia);
+        dorms.add(Alvey);
+        dorms.add(Bushnell);
+        dorms.add(Jefferson);
+        dorms.add(Randolph);
+        dorms.add(Russell);
+    }
+
+    /**
+     * Singleton pattern.
+     */
+    public static synchronized FreshmanHousingSelection instance(){
+        if(theInstance==null){
+            theInstance=new FreshmanHousingSelection();
+        }
+        return theInstance;
+    }
+
+    /** 
+     * Assign the set of (presumably freshman, but this is not checked)
+     * students passed to their freshman dorm rooms, in a way that does not
+     * take race into account. (compare {@link #assignByRace}.) As a precursor
+     * (side effect) to this, any freshmen already existing in freshmen dorms
+     * will be removed.
+     */
+    public void assign(Bag newFreshmen) {
+        emptyDorms();
+        for(int x=0;x<newFreshmen.size();x++){
+            Student s = (Student) newFreshmen.get(x);
+            for(int y=0; y<dorms.size();y++){
+                if(s.hasRoom()){break;}
+                Dorm d = (Dorm) dorms.get(y);
+                if(d.isFull()){
+                    continue;
+                }
+                if(d.isFemaleOnly() && s.getGender()==Student.Gender.FEMALE){
+                    continue;
+                }
+                int i = 0;
+                while(i<d.getNumRooms()){
+                    if(s.hasRoom()){break;}
+                    try{
+                        d.getRoomByIndex(i).addResident(s);
+                    }catch(IllegalArgumentException e){//System.out.println(e);
+                    }
+                    i++;
+
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 60:79, 182:202.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.